src/activities/home/GridBrowserActivity.cpp

```diff
@@ -0,0 +1,309 @@
+      // open subfolder
+      basepath += files[selected].name;
+      loadFiles();
+      onPageChanged();
+    } else {
+      onSelect(basepath + files[selected].name);
+    }
+  } else if (inputManager.wasPressed(InputManager::BTN_BACK)) {
+    if (basepath != "/") {
+      basepath.resize(basepath.rfind('/'));
+      if (basepath.empty()) basepath = "/";
+      loadFiles();
+      onPageChanged();
+    } else {
+      // At root level, go back home
+      onGoHome();
+    }
+  } else if (prevReleased) {
+    previousSelectorIndex = selectorIndex;
+    if (selectorIndex == 0 || skipPage) {
+      if (page > 0) {
+        page--;
+        onPageChanged();
+      }
+    } else {
+      selectorIndex--;
+      updateRequired = true;
+    }
+  } else if (nextReleased) {
+    previousSelectorIndex = selectorIndex;
+    if (selectorIndex == min(PAGE_ITEMS, files.size() - page * PAGE_ITEMS) - 1 || skipPage) {
+      if (page < files.size() / PAGE_ITEMS) {
+        page++;
+        onPageChanged();
+      }
+    } else {
+      selectorIndex++;
+      updateRequired = true;
+    }
+  }
+}
+
+void GridBrowserActivity::displayTaskLoop() {
+  while (true) {
+    if (renderRequired || updateRequired) {
+      bool didRequireRender = renderRequired;
+      renderRequired = false;
+      updateRequired = false;
+      xSemaphoreTake(renderingMutex, portMAX_DELAY);
+      render(didRequireRender);
+      xSemaphoreGive(renderingMutex);
+    }
+    vTaskDelay(10 / portTICK_PERIOD_MS);
+  }
+}
+
+void GridBrowserActivity::render(bool clear) const {
+  if (clear) {
+    renderer.clearScreen();
+    auto folderName = basepath == "/" ? "SD card" : basepath.substr(basepath.rfind('/') + 1).c_str();
+    drawFullscreenWindowFrame(renderer, folderName);
+  }
+  
+  if (!files.empty()) {      
+    for (size_t i = 0; i < min(PAGE_ITEMS, files.size() - page * PAGE_ITEMS); i++) {
+      const auto file = files[i + page * PAGE_ITEMS];
+      
+      const int16_t tileX = GRID_OFFSET_LEFT + i % 3 * TILE_W;
+      const int16_t tileY = GRID_OFFSET_TOP + i / 3 * TILE_H;
+
+      if (file.type == F_DIRECTORY) {
+        constexpr int iconOffsetX = (TILE_W - FOLDERICON_WIDTH) / 2;
+        constexpr int iconOffsetY = (TILE_H - TILE_TEXT_H - FOLDERICON_HEIGHT) / 2;
+        renderer.drawIcon(FolderIcon, tileX + iconOffsetX, tileY + iconOffsetY, FOLDERICON_WIDTH, FOLDERICON_HEIGHT);
+      }
+
+      if (!file.thumbPath.empty()) {
+        Serial.printf("Rendering file thumb: %s\n", file.thumbPath.c_str());
+        File bmpFile = SD.open(file.thumbPath.c_str());
+        if (bmpFile) {
+          Bitmap bitmap(bmpFile);
+          if (bitmap.parseHeaders() == BmpReaderError::Ok) {
+            constexpr int thumbOffsetX = (TILE_W - THUMB_W) / 2;
+            constexpr int thumbOffsetY = (TILE_H - TILE_TEXT_H - THUMB_H) / 2;
+            renderer.drawBitmap(bitmap, tileX + thumbOffsetX, tileY + thumbOffsetY, THUMB_W, THUMB_H);
+          }
+        }
+      }
```

Comment:
Would be nice to have a fallback icon for books where the cover can't be generated